### PR TITLE
3/upload noties

### PR DESCRIPTION
**Pull request by abea**
Resolves https://apostrophecms.atlassian.net/browse/A30U-393

This delivers notifications to the end user for uploading images:
- starting an upload
- successful upload
- error from uploading an unacceptable file type (one apostrophe doesn't know, e.g., json)
- error from uploading an incorrect file type for that field (e.g., pdf)
- general error messages if there's not something more specific.

### CODE REVIEW

File: modules/@apostrophecms/attachment/index.js [hunk 1]
```diff
@@ -285,9 +285,18 @@ module.exports = {
           return;
         }
         _.assign(info, _.omit(dbInfo, 'crop'));
-        if (!self.acceptableExtension(field, info)) {
-          throw self.apos.error('invalid');
+
+        // Check if the file type is acceptable of if there are
+        const correctedExtensions = self.checkExtension(field, info);
```
Comment (abea) [inline modules/@apostrophecms/attachment/index.js:290]: I wanted to get the list of acceptable extensions from this, so I changed how this works.

File: modules/@apostrophecms/attachment/index.js [hunk 2]
```diff
@@ -308,29 +317,34 @@ module.exports = {
...
+      checkExtension(field, attachment) {
+        const groups = field.fileGroups ||
+          (field.fileGroup && [ field.fileGroup ]);
         let extensions;
+
         if (groups) {
           if (!_.includes(groups, attachment.group)) {
             extensions = [];
             _.each(groups, function (group) {
-              const groupInfo = _.find(self.options.fileGroups, { name: group });
```
Comment (abea) [inline modules/@apostrophecms/attachment/index.js:318]: Fix here. Please confirm.

File: modules/@apostrophecms/attachment/index.js [hunk 3]
```diff
@@ -308,29 +317,34 @@ module.exports = {
...
               extensions = extensions.concat(groupInfo.extensions);
             });
-            return false;
+
+            return extensions;
           }
         }
-        extensions = field.extensions || (field.extension && [ field.extension ]);
+        extensions = field.extensions ||
+          (attachment.extension && [ attachment.extension ]);
```
Comment (abea) [inline modules/@apostrophecms/attachment/index.js:340]: Fix here. Please confirm.

File: modules/@apostrophecms/schema/index.js [hunk 4]
```diff
@@ -1427,6 +1425,9 @@ module.exports = {
             try {
               await convert(req, field, data, object);
             } catch (e) {
+              if (e instanceof Error) {
```
Comment (abea) [inline modules/@apostrophecms/schema/index.js:1428]: If we get an `Error` send it on.

File: modules/@apostrophecms/attachment/index.js (hunk 2 above)
Comment (boutell) [reply modules/@apostrophecms/attachment/index.js:318]: 👍

File: modules/@apostrophecms/attachment/index.js (hunk 3 above)
Comment (boutell) [reply modules/@apostrophecms/attachment/index.js:340]: attachment.extension has no business on a list of what's allowed; it's what was actually sent.

The idea here was to support attachment fields specifying a single extension to allow or more than one, as an alternative to specifying a group. The existing code seems correct for that.

Comment (boutell) [reply modules/@apostrophecms/attachment/index.js:340]: A method that returns either `true` or an array of things that would have been OK is unusual, so please document it.

File: modules/@apostrophecms/attachment/index.js [hunk 5]
```diff
@@ -947,7 +964,10 @@ module.exports = {
       canUpload(req, res, next) {
         if (!self.apos.permission.can(req, 'edit-attachment')) {
           res.statusCode = 403;
-          return res.send('forbidden');
+          return res.send({
+            type: 'forbidden',
+            message: req.__('You do not have enough permissions to upload')
```
Comment (boutell) [inline modules/@apostrophecms/attachment/index.js:969]: You do not have permission to upload a file.

Comment (boutell) [reply modules/@apostrophecms/attachment/index.js:969]: "Enough permissions" seems needlessly technical.

File: modules/@apostrophecms/image/ui/apos/components/AposMediaManagerDisplay.vue [hunk 6]
```diff
@@ -132,13 +132,21 @@ export default {
...
       const file = event.target.files[0];
 
       const emptyDoc = await apos.http.post(this.moduleOptions.action, {
         body: {
           _newInstance: true
         }
       });
+      await apos.notify(
+        // TODO: i18n
+        `Uploading ${fileCount} image${fileCount > 1 ? 's' : ''}`,
```
Comment (boutell) [inline modules/@apostrophecms/image/ui/apos/components/AposMediaManagerDisplay.vue:145]: I was about to tell you how to do this with %s placeholders and more arguments to apos.notify, but it involves pluralization and we don't actually have a wrapper for that yet. Hmm. OK for now.

File: modules/@apostrophecms/notification/ui/apos/apps/ApostropheNotification.js [hunk 7]
```diff
@@ -121,6 +122,7 @@ export default function() {
           :key="notification._id"
           :label="notification.message"
           :type="notification.type"
+          :icon="notification.icon"
```
Comment (boutell) [inline modules/@apostrophecms/notification/ui/apos/apps/ApostropheNotification.js:125]: Do we need icon in addition to type? Should we just map types to icons or something? Is this part of Stu's design?

File: modules/@apostrophecms/schema/index.js (hunk 4 above)
Comment (boutell) [reply modules/@apostrophecms/schema/index.js:1428]: This would send on pretty much everything, because self.apos.error does construct an Error.

Comment (boutell) [reply modules/@apostrophecms/schema/index.js:1428]: Not sure what you're trying to do.

**Review by boutell — CHANGES_REQUESTED** [review]

File: modules/@apostrophecms/attachment/index.js (hunk 3 above)
Comment (abea) [reply modules/@apostrophecms/attachment/index.js:340]: True. For some reason that seemed necessary, but the `self.options.fileGroups` fix was probably the important one.

File: modules/@apostrophecms/notification/ui/apos/apps/ApostropheNotification.js (hunk 7 above)
Comment (abea) [reply modules/@apostrophecms/notification/ui/apos/apps/ApostropheNotification.js:125]: they are passed in separately right now. The current implementation increases flexibility, but I'd be happy for it to be more limited to simplify the implementation. Out of scope of this PR, though.

File: modules/@apostrophecms/schema/index.js (hunk 4 above)
Comment (abea) [reply modules/@apostrophecms/schema/index.js]: The problem was that an error thrown in the field's `convert` method was buried here, including the helpful message. I couldn't see the `message` property on it by logging, but I see it's there now to check and use.

File: modules/@apostrophecms/schema/index.js [hunk 8]
```diff
@@ -1425,8 +1425,8 @@ module.exports = {
             try {
               await convert(req, field, data, object);
             } catch (e) {
-              if (e instanceof Error) {
-                throw e;
+              if (e instanceof Error && e.message) {
```
Comment (boutell) [inline modules/@apostrophecms/schema/index.js:1428]: This isn't going to work either, but it's my fault. The error module doesn't currently set any property that doesn't exist on a generic Error, or potentially other people's third party subclasses of Error. We need to duck type this.

In the error module, please set `.aposError = true` on errors constructed by the error module. Then you can test for it here successfully. We could also include an `isOurs` method that tests it, for calling `apos.error.isOurs(error)`.

Let me know if you'd rather I did a separate PR for that first but I'm trying to speed you along.

**Review by boutell — CHANGES_REQUESTED** [review]

**Review by boutell — APPROVED** [review]

**Review by falkodev — APPROVED** [review]

**Review by falkodev — APPROVED** [review]

**Review by boutell — APPROVED** [review]

**Pull request merged**